applet/src/com/redhat/ckey/applet/CardEdge.java

```diff
@@ -538,22 +561,37 @@ private CardEdge(byte bArray[], short bOffset, byte bLength)
         ciphers       = new Cipher    [MAX_NUM_KEYS];
         signatures    = new Signature [MAX_NUM_KEYS];
         default_nonce = new byte      [NONCE_SIZE];
+        default_iv    = new byte      [NONCE_SIZE *2];
         issuerInfo    = new byte      [ISSUER_INFO_SIZE];
 
+
+	/* AES KWP allocations
+	 *
+	 */
+
+        kwp_t = new Unsigned64();
+        kwp_B0 = new Unsigned64();
+        kwp_B1 = new Unsigned64();
+        kwp_R  = new Unsigned64();
+        kwp_iv64 = new Unsigned64();
```

Comment:
I'm a little worried about these. This allocates The Unsigned64 ints out of persistent memory. That means the data point is in persistent memory (the pointer, not the data it points to). Every time you do a setData() it will set your data pointer, wearing down that memory. i think you need to allocate these our of transient memory.